agent/hcp/client/metrics_client.go

```diff
@@ -0,0 +1,158 @@
+}
+
+// TelemetryClientCfg is used to configure the MetricsClient.
+type TelemetryClientCfg struct {
+	CloudCfg cloudConfig
+	Logger   hclog.Logger
+}
+
+// NewMetricsClient returns a configured MetricsClient.
+// The current implementation uses otlpClient to provide retry functionality.
+func NewMetricsClient(cfg *TelemetryClientCfg) (MetricsClient, error) {
+	if cfg.CloudCfg == nil || cfg.Logger == nil {
+		return nil, fmt.Errorf("failed to init telemetry client: provide valid TelemetryClientCfg")
+	}
+
+	c, err := newHTTPClient(cfg.CloudCfg, cfg.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init telemetry client: %v", err)
+	}
+
+	headers := map[string]string{
+		"X-HCP-Source-Channel": fmt.Sprintf("consul %s hcp-go-sdk/%s", version.GetHumanVersion(), version.Version),
+		"Content-Type":         "application/x-protobuf",
+	}
+
+	return &otlpClient{
+		client:  c,
+		headers: headers,
+	}, nil
+}
+
+// newHTTPClient configures the retryable HTTP client.
+func newHTTPClient(cloudCfg cloudConfig, logger hclog.Logger) (*retryablehttp.Client, error) {
+	hcpCfg, err := cloudCfg.HCPConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	tlsTransport := cleanhttp.DefaultPooledTransport()
+	tlsTransport.TLSClientConfig = hcpCfg.APITLSConfig()
+
+	var transport http.RoundTripper = &oauth2.Transport{
+		Base:   tlsTransport,
+		Source: hcpCfg,
+	}
+
+	client := &http.Client{
+		Transport: transport,
+		Timeout:   defaultStreamTimeout,
+	}
+
+	retryClient := &retryablehttp.Client{
+		HTTPClient:   client,
+		Logger:       logger.Named("hcp_telemetry_client"),
+		RetryWaitMin: defaultRetryWaitMin,
+		RetryWaitMax: defaultRetryWaitMax,
+		RetryMax:     defaultRetryMax,
+		CheckRetry:   retryablehttp.DefaultRetryPolicy,
+		Backoff:      retryablehttp.DefaultBackoff,
+	}
+
+	return retryClient, nil
+}
+
+// ExportMetrics is the single method exposed by MetricsClient to export OTLP metrics to the desired HCP endpoint.
+// The endpoint is configurable as the endpoint can change during periodic refresh of CCM telemetry config.
+// By configuring the endpoint here, we can re-use the same client and override the endpoint when making a request.
+func (o *otlpClient) ExportMetrics(ctx context.Context, protoMetrics *metricpb.ResourceMetrics, endpoint string) error {
+	pbRequest := &colmetricpb.ExportMetricsServiceRequest{
+		ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
+	}
+
+	body, err := proto.Marshal(pbRequest)
+	if err != nil {
+		return fmt.Errorf("failed to export metrics: %v", err)
+	}
+
+	req, err := retryablehttp.NewRequest(http.MethodPost, endpoint, bytes.NewBuffer(body))
+	if err != nil {
+		return fmt.Errorf("failed to export metrics: %v", err)
+	}
+
+	for k, v := range o.headers {
+		req.Header.Set(k, v)
+	}
+
+	resp, err := o.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to export metrics: %v", err)
+	}
+
+	var respData bytes.Buffer
+	if _, err := io.Copy(&respData, resp.Body); err != nil {
```

Comment:
I can't remember if `io.Copy` does this for us (I don't think so), but we should probably call `defer resp.Body.Close()` beforehand.

Reply:
weird.. seems like that got deleted in one of my rebases. Was in a previous commit. Good catch, we defo need to close it.

https://github.com/hashicorp/consul/pull/17100/commits/b2b44f5db957e3c7bf8be7f7fc984b445b3bdc09